### DISCUSSION TITLE
How To Use Linux To Build/Boot the DCC/EX-T3-WIFI Throttle Code for The FireBeetle2 ESP32-WROOM-32E

### DISCUSSION THREAD

**Discussion by daveclifford** [DCC-EX/EX-T3-WiFi#43] [General]
I was able to use my Linux Development system to build/upload the build to the FireBeetle2. I hope this helps other Linux Developers,

The steps/issues I encountered during the support software installation and firmware build/load execution were:

    1. Ensuring that the USB-C to USB-A (or USB-C) Interface Cable is Both a Power AND Data Cable 
    2. Installing The VisualStudio Code (VSCode) and Platform IO (PlatformIO) IDE Code
    3. Installing the EX-T3-WiFi Git Repository
    4. TTY ports needed to be enabled for use
    5. Fix for Linux /dev/ttyUSB0  - the Port Doesn't Exist 
    6. FireBeetle2 BootLoader mode needed attention 

Use Linux To Build-Boot the DCC-EX-T3-WIFI Throttle Code .odt

(https://github.com/DCC-EX/EX-T3-WiFi/files/10532847/Use.Linux.To.Build-Boot.the.DCC-EX-T3-WIFI.Throttle.Code.odt)

**Comment by daveclifford** [comment]
I am going to add to the document that the upload **only works one time** after the FireBeetle2 is plugged in. If I  attempt to upload a second time, the following error occurs:  

> A fatal error occurred: Failed to connect to ESP32: Wrong boot mode detected (0x13)! The chip needs to be in download 
> mode.

I assume the the PIO or FireBeetle2  is initially reset to download mode,, and, after the upload occurs, is set to a different state.

The solution might be to do some magic with an upload directive.

I confirmed that this happens by trying to upload two times in a row **FAILS**, and then disconnecting, reconnecting the FireBeetle2 and uploading successfully.

Dave

**Reply by alex-code** [reply]
I've never had to manually reconnect to upload again.
Did you have a serial output terminal open?

I've had it before where if I did an **Upload and Monitor** I'd have to close that terminal before I could upload the file system as the port was in use.

**Reply by alex-code** [reply]
Seems like there's an issue with the uploader on Mac and Linux,
https://github.com/espressif/esptool/issues/712 - https://github.com/espressif/esptool/issues/741

Some of the comments mention using a USB2.0 port and it working.

**Reply by daveclifford** [reply]
Thanks Alex!  Saw this in a MicroRobotics ad "**FireBeetle-ESP32 adopts CH340 serial chip** that can be used without driver among most devices. If you find the driver is not installed automatically after plugging into the device, you can install it manually; [CH340 Driver](https://dfimg.dfrobot.com/nobody/wiki/0e0d6b3864f7163833ec5d7ad4af7632.EXE)" I'm not surprised that Linux support is lacking compared to MSWIN. I got the /dev/USB0 interface working (CH340 driver installed, recognized, and allowed to be used by logged-in user). Is there a way in ProgrammedIO where, when the UPLOAD task is executed, I could add some code to use **esptool.py** to reset the port to download?   There exists a sequence to reset the port to download. 

I will look at your references and investigate further. Thanks again. 

(I am going to spend some time today soldering the connectors to the Rotary Encoder and Accelerometer - 5 degrees F)

**Comment by daveclifford** [comment]
I think that the ESP32 CH340 serial emulation code has  timing issues.
(maybe - just a guess)

I am stubborn enough to find a s/w solution.

My plan is to modify the esptools.py  code to provide the delay
consistently, and  then, with your help,  call the code before the upload
occurs in programmedIO.

Dave
